README.md

![BADASS logo](https://github.com/remingtonsexton/BADASS3/blob/master/figures/BADASS2_logo.gif)

## New in Version 10.2.0
* New line and configuration testing framework. See [Line Testing and Options](#line-testing-and-options).
* Line component options.  See [Line Component Options](#line-component-options).
* Improvements in autocorrelation calculations.
* W80 now a standard output parameter for all lines.
* Outputs line widths that are both corrected and uncorrected for input resolution
* BADASS now normalizes the spectrum internally for fitting purposes.
* Various improvements to global optimizer used in initial fit.
* **Note**: The algorithm used for scaling the noise to achieve a $\Chi_\nu^2=1$ was found to be numerically unstable and users should use `fit_stat='ML'` instead.
<hr>

Ridiculous acronyms are a [long-running joke in astronomy](https://lweb.cfa.harvard.edu/~gpetitpas/Links/Astroacro.html), but here, spectral fitting ain't no joke!

[BADASS](https://ui.adsabs.harvard.edu/abs/2021MNRAS.500.2871S/abstract) is an open-source spectral analysis tool designed for detailed decomposition of Sloan Digital Sky Survey (SDSS) spectra, and specifically designed for the fitting of Type 1 ("broad line") Active Galactic Nuclei (AGN) in the optical.  The fitting process utilizes the Bayesian affine-invariant Markov-Chain Monte Carlo sampler [emcee](https://ui.adsabs.harvard.edu/abs/2013PASP..125..306F/abstract) for robust parameter and uncertainty estimation, as well as autocorrelation analysis to access parameter chain convergence.  BADASS can fit the following spectral features:
- Stellar line-of-sight velocity distribution (LOSVD) using Penalized Pixel-Fitting ([pPXF](https://www-astro.physics.ox.ac.uk/~mxc/software/#ppxf), [Cappellari et al. (2017)](https://ui.adsabs.harvard.edu/abs/2017MNRAS.466..798C/abstract)) using templates from the [Indo-U.S. Library of Coudé Feed Stellar Spectra](https://www.noao.edu/cflib/) ([Valdes et al. (2004)](https://ui.adsabs.harvard.edu/abs/2004ApJS..152..251V/abstract)) in the optical region 3460 Å - 9464 Å.
- Broad and Narrow FeII emission features using the FeII templates from [Véron-Cetty et al. (2004)](https://ui.adsabs.harvard.edu/abs/2004A%26A...417..515V/abstract) or [Kovačević et al. (2010)](https://ui.adsabs.harvard.edu/abs/2010MSAIS..15..176K/abstract).
- UV iron template from [Vestergaard and Wilkes (2001)](https://ui.adsabs.harvard.edu/abs/2001ApJS..134....1V/abstract)
- Individual narrow, broad, and/or absorption line features.
- AGN power-law continuum and Balmer pseudo-continuum.
- "Blue-wing" outflow emission components found in narrow-line emission. 

A more-detailed summary of BADASS, as well as a case-study of ionized gas outflows, is given in [Sexton et al. (2021)](https://ui.adsabs.harvard.edu/abs/2021MNRAS.500.2871S/abstract). 

All spectral components can be turned off and on via the [Jupyter Notebook](https://jupyter.org/) interface, from which all fitting options can be easily changed to fit non-AGN-host galaxies (or even stars!).  BADASS uses multiprocessing to fit multiple spectra simultaneously depending on your hardware configuration.  The code was originally written in Python 2.7 to fit Keck Low-Resolution Imaging Spectrometer (LRIS) data ([Sexton et al. (2019)](https://ui.adsabs.harvard.edu/abs/2019ApJ...878..101S/abstract)), but because BADASS is open-source and *not* written in an expensive proprietary language, one can easily contribute to or modify the code to fit data from other instruments.  Out of the box, BADASS fits SDSS spectra, MANGA IFU cube data, and examples are provided for fitting user-input spectra of any instrument.

Before getting started you should read the readme below.

## Table of Contents
- [Installation](#installation)
- [Usage](#usage)
  * [Fitting Options](#fitting-options)
  * [MCMC Options](#mcmc-options)
  * [Model Component Options](#model-component-options)
  * [Multiple Line Components](#multiple-line-components)
  * [Line Testing and Options](#line-testing-and-options)
  * [Configuration Testing](#configuration-testing)
  * [Line Component Options](#line-component-options)
  * [User Lines, Constraints, and Masks](#user-lines-constraints-and-masks)
  * [Combined Lines](#combined-lines)
  * [LOSVD Fitting Options (pPXF)](#losvd-fitting-options-ppxf)
  * [Host Model Options](#host-model-options)
  * [Power Law Options](#power-law-options)
  * [Polynomial Options](#polynomial-options)
  * [Optical FeII Options](#optical-feii-options)
  * [UV Iron Options](#uv-iron-options)
  * [Balmer Pseudo-Continuum Options](#balmer-psuedocontinuum-options)
  * [Plotting Options](#plotting-options)
  * [Output Options](#output-options)
  * [The Main Function and Calling Sequence](#the-main-function-and-calling-sequence)
- [Output](#output)
  * [Best-fit Model](#best-fit-model)
  * [Parameter Chains, Histograms, Best-fit Values and Uncertainties](#parameter-chains-histograms-best-fit-values-and-uncertainties)
  * [Log File](#log-file)
  * [Best-fit Model Components](#best-fit-model-components)
  * [Best-fit Parameters and Uncertainties](#best-fit-parameters-and-uncertainties)
  * [Autocorrelation Analysis](#autocorrelation-analysis)
- [Examples & Tutorials](#examples)
  * [Basic Example: A Single SDSS Spectrum](#single-sdss-spectrum)
  * [A Single Non-SDSS Spectrum](#single-non-sdss-spectrum)
  * [Multiple Spectra with Multiprocessing](#multiple-spectra-with-multiprocessing)
  * [MANGA IFU Cube Data](#manga-ifu-cube-data)
  * [Non-MANGA IFU Cube Data](#non-manga-ifu-cube-data)
- [How to...](#how-to)
  * [Prior Distributions](#prior-distributions)
  * [Line Lists](#line-lists)
  * [Hard Constraints](#hard-constraints)
  * [Soft Constraints](#soft-constraints)
- [Known Issues](#known-issues)
- [Credits](#credits)
- [License](#license)



# Installation

The easiest way to get started is to simply clone the repository. 

As of the most recent version, the following packages are required (Python 3.8.12):

- [`numpy 1.22.3`](https://numpy.org/doc/stable/index.html)
- [`pandas 1.4.2`](https://pandas.pydata.org/)
- [`scipy 1.7.3` ](https://scipy.org/)
- [`matplotlib 3.7.1`](https://matplotlib.org/)
- [`astropy 5.2.1`](https://www.astropy.org/)
- [`astroquery 0.4.6`](https://astroquery.readthedocs.io/en/latest/)
- [`emcee 3.1.4`](https://emcee.readthedocs.io/en/stable/)
- [`numexpr 2.8.4`](https://github.com/pydata/numexpr)
- [`natsort 8.3.1`](https://natsort.readthedocs.io/en/master/)
- [`psutil 5.9.4`](https://psutil.readthedocs.io/en/latest/)
- [`vorbin 3.1.5`](https://www-astro.physics.ox.ac.uk/~cappellari/software/#binning)
<!-- - [`astro-bifrost 2.0.3`](https://pypi.org/project/astro-bifrost/) -->
- [`spectres 2.2.0`](https://spectres.readthedocs.io/en/latest/)
- [`corner 2.2.1`](https://corner.readthedocs.io/en/latest/)
- [`prettytable 3.6.0`](https://pypi.org/project/prettytable/)
- **Note**: [`ppxf`](https://www-astro.physics.ox.ac.uk/~cappellari/software/#ppxf) was integrated within the BADASS source code early on, and does not require installation.
- **Optional**: [`plotly 5.13.1`](https://plotly.com/python/getting-started/) (for interactive HTML plots)

The code is run entirely through the Jupyter Notebook interface, and is set up to run on the included spectrum files in the ".../examples/" folder.  If one wants to fit multiple spectra consecutively, simply add folders for each spectrum to the folder.  This is the recommended directory structure:

![directory structure](https://github.com/remingtonsexton/BADASS3/blob/master/figures/BADASS_example_spectra.png)

Simply create a folder containing the SDSS FITS format spectrum, and BADASS will generate a working folder for each fit titled "MCMC_output_#" inside each spectrum folder.  BADASS automatically generates a new output folder if the same object is fit again (it does not delete previous fits).

# Usage

## Fitting Options
 
```python 
################################## Fit Options #################################
# Fitting Parameters
fit_options={
fit_options={
"fit_reg"    : (4400,5500),# Fitting region; Note: Indo-US Library=(3460,9464)
"good_thresh": 0.0, # percentage of "good" pixels required in fig_reg for fit.
"mask_bad_pix": False, # mask pixels SDSS flagged as 'bad' (careful!)
"mask_emline" : False, # automatically mask lines for continuum fitting.
"mask_metal": False, # interpolate over metal absorption lines for high-z spectra
"fit_stat": "RCHI2", # fit statistic; ML = Max. Like. , OLS = Ordinary Least Squares, RCHI2 = reduced chi2
"n_basinhop": 15, # Number of consecutive basinhopping thresholds before solution achieved
"test_lines": False, # Perform line/configuration testing for multiple components
"max_like_niter": 10, # number of maximum likelihood iterations
"output_pars": False, # only output free parameters of fit and stop code (diagnostic)
"cosmology": {"H0":70.0, "Om0": 0.30}, # Flat Lam-CDM Cosmology
}
################################################################################
```

**`fit_reg`**: (*tuple/list of length (2,)*; *Default: (4400,5500)*)<br/>
the minimum and maximum desired fitting wavelength in angstroms, for example (4400,7000).  This is passed to the `determine_fit_reg()` function to check if this region is valid and and which emission lines to fit.

**`good_thresh`**: (*float [0.0,1.0]*; *Default: 0.0*)<br/>
the cutoff for minimum fraction of "good" pixels (determined by SDSS) within the fitting range to allow for fitting of a given spectrum.  If the spectrum has fewer good pixels than this value, BADASS skips over it and moves onto the next spectrum.

**`mask_bad_pix`**: (*bool*; *Default: False*)<br/>
Mask pixels which SDSS flagged as bad due to sky line subtraction or cosmic rays.  Warning: if large portions of the fitting region are marked as bad pixels, this can cause BADASS to crash.  One should only use this if only a few pixels are affected by contamination.  

**`mask_emline`**: (*bool*; *Default: False*)<br/>
Mask any significant absorption and emission features relative to the continuum.  This uses an automated iterative moving median filter of various sizes to detect significant flux differences between window sizes.  Good for continuum fitting but tends to over mask lots of features near the edges of the spectrum.

**`mask_metal`**: (*bool*; *Default: False*)<br/>
Performs the same moving median filter algorithm as `mask_emline` but only to absorption features.  Works well for metal absorption features seen typically in high-redshift spectra.

**`fit_stat`**: (*str*; *Default: : "ML"*)<br/>
The fit statistic used for the likelihood.  The default is "RCHI2" which converges on a reduced chi-squared of 1 by scaling the input noise by a `noise_scale` free parameter.  Other options include "ML" for standard maximum likelihood (pixels weighted by noise with no noise scaling), and "LS" for ordinary least-squares fitting (all pixels weighted by same amount).

**`n_basinhop`**: (*int*; *Default: 25*)<br/>
Number of successive `niter_success` times the basinhopping algorithm needs to achieve a solution.  The fit becomes much better with more success times, however this can increase the time to a solution significantly  Recommended 5-10. 

**`reweighting`**: (*bool*; *Default: True*)<br/>
If true, BADASS will reweight the noise vector to achieve a reduced chi-squared ~ 1.  This is done after the initial basinhopping fit, and applied to any bootstrapped uncertainties and MCMC fitting performed afterward.  This does not affect the chi-squared ratio metric used in line and configuration testing, but does effect the amplitude-over-noise and SNR calculations in BADASS.

**`test_lines`**: (`bool`:*bool*; *Default: False*)<br/>
Performs tests for lines with multiple components (parent-child pairs).  Options are specified in `test_options`.

**`max_like_niter`**: (*int*; *Default: 10*)<br/>
Number of bootstrapping iterations to perform after the initial basinhopping fit.  This is a means to obtain uncertainties on parameters without performing MCMC fitting, however, do not produce as robust uncertainties as MCMC.

**`output_pars`**: (*bool*; *Default: False*)  <br/>
Convenience feature that prints out all free parameters so the user can check and then terminates without fitting.

**`cosmology`**: (*Default*: `{"H0":70.0, "Om0": 0.30}`)<br/>
The flat Lambda-CDM cosmology assumed for calculating luminosities from fluxes. 


# MCMC Options

```python
########################### MCMC algorithm parameters ##########################
mcmc_options={
  "mcmc_fit"    : False, # Perform robust fitting using emcee
  "nwalkers"    : 100,  # Number of emcee walkers; min = 2 x N_parameters
  "auto_stop"   : False, # Automatic stop using autocorrelation analysis
  "conv_type"   : "all", # "median", "mean", "all", or (tuple) of parameters
  "min_samp"    : 1000,  # min number of iterations for sampling post-convergence
  "ncor_times"  : 1.0,  # number of autocorrelation times for convergence
  "autocorr_tol": 10.0,  # percent tolerance between checking autocorr. times
  "write_iter"  : 100,   # write/check autocorrelation times interval
  "write_thresh": 100,   # iteration to start writing/checking parameters
  "burn_in"     : 1500, # burn-in if max_iter is reached
  "min_iter"    : 1000, # min number of iterations before stopping
  "max_iter"    : 2500, # max number of MCMC iterations
}
################################################################################
```

**`mcmc_fit`**: (*bool*; *Default: False*) <br/>
Perform fit with MCMC using the initial maximum likelihood fit as initial parameters for the fit.  It is *highly recommended* that one use MCMC to perform the fit, although sampling will require a significant amount of time compared to a maximum likelihood fit using `scipy.optimize.minimize()`.

**`nwalkers`**: (*int*; *Default: 100*) <br/> 
number of "walkers" per parameter used by emcee to explore each parameter space.  The minimum number of walkers is 2 x ( # of free parameters), set by emcee.

**`auto_stop`**: (*bool*; *Default: True*) <br/> 
if set to *True*, autocorrelation is used to automatically stop the fitting process when a convergence criteria (`conv_type`) is achieved. 

**`conv_type`**: (*str*; *Default: "median"*; *options: "all", "median", "mean", list of parameters*)  
mode of convergence.  Convergence of 'all' ensures all fitting parameters have achieved the desired `ncor_times` and `autocorr_tol` criteria, while "median" and "mean" only ensure that `ncor_times` and `autocorr_tol` criteria have been met for the median or mean of all parameters, respectively.  A list of valid parameters is also acceptable to ensure specific parameters have achieved convergence even if others have not.  In general "median" requires the fewest number of iterations and is not sensitive to poorly-constrained parameters, and "all" and "mean" require the most number of iterations and are much more sensitive to fluctuations in calculated autocorrelation times and tolerances.  A list of parameters is suitable in cases where one is only interested in certain spectral features.

**`min_samp`**: (*int*; *Default: 1000*)  <br/>
if `auto_stop=True`, then the `burn_in` is the iteration at which convergence is achieved, and `min_samp` is the number of iterations *past convergence* used for posterior sampling (the samples used for histograms and estimating best-fit parameters and uncertainties.  If for some reason the parameters "jump out" of convergence, the `burn_in` will reset and BADASS will continue to sample until convergence is met again.  If emcee completes `min_samp` iterations after convergence is achieved without jumping out of convergence, this concludes the MCMC sampling.

**`ncor_times`**: (*int* or *float*; *Default=10*) <br/> 
The number of integrated autocorrelation times (iterations) needed for convergence.  We recommend a minimum of `ncor_times=2.0`.  In general, it will require more than 2.0 autocorrelation times to calculate the autocorrelation time for a parameter chain.  Increasing `ncor_times` ensures that the parameter chain has stopped exploring the parameter space and is ready to begin sampling for the posterior distribution. 

**`autocorr_tol`**: (*int* or *float*; *Default=10*) <br/>

CHANGES.rst
10.0.0 - 10.3.0
===============
- New generalized line component option for easily adding n number of line components; deprecates 'outflow' components.
- W80 now a standard line parameter.
- New testing suite that incorporates testing for multiple components, lines, metrics, etc. with the 
  ability to continue fitting with the best model.
- Fixed autocorrelation time calculation; now always produces a time.
- PPOLY polynomial no longer an option pending bug fixes.
- To avoid an excessive number of plots, we now limit plotting of histograms to free fitted parameters.
- Configuration testing.
- Removed astro-bifrost due to grpcio dependency problems; using old metal masking algorithm until fixed.
- BADASS now internally normalizes the spectrum for optimization reasons.
- Bug fixes and minor changes.
- Reweighting noise to achieve RCHI2=1 now done prior to bootstrapping and MCMC as needed (no longer a free parameter, which made it numerically unstable).


9.3.1
=====
- BADASS-IFU
	- S/R computed at 5100 angstroms (rest frame) by default for use when using Voronoi binning
- Bug fixes, and edits to default line list
- Add explicit flat prior
- Add flux normalization option (default is SDSS normalization of 1.E-17)
- Fixed output line SNR to be calculated even if NPIX <1
- Constraint and initial value checking before fit takes place to prevent crashing
- Implemented restart file; saves all fitting options to restart fit


9.3.0
=====
- NPIX and SNR (signal-to-noise ratio) is computed for all lines and now includes an uncertainty.
- Removed interpolation inside of fit_model() to reduce computational expense.
- General bug fixes and cleaning up.


9.2.0 - 9.2.2
=============
- Options for different priors on free parameters
- Normalization for log-likelihoods
- Outflow test region fix


9.1.7
=====
- Switched width parameter of all lines from FWHM to dispersion to accomodate more lines and avert problems with biased integrated velocities and dispersions. As a result, integrated dispersions and velocities are only calculated for combined lines, and FWHM are calculated for ALL lines.
- Added Laplace and Uniform line profiles from Sanders et al. (2020) (https://ui.adsabs.harvard.edu/abs/2020MNRAS.499.5806S/abstract; https://github.com/jls713/gh_alternative)
- Changed instrumental fwhm keyword to instrumental dispersion "disp_res".  Input resolution for user input spectra is still a "fwhm_res" but changes to disp_res internally.


9.1.6
=====
- Polynomial continuum components independent from LOSVD component
- Linearization of non-linearized non-SDSS spectra using spectres module


9.0.0 - 9.1.1
=============
- Options for likelihood function
- Consolidated outflow and line testing routines


8.0.14 - 8.0.15
===============
- Regular expressions now supported for soft constraints
- IFU support for MANGA and MUSE (General) datasets


8.0.0 - 8.0.13
==============
- Added smoothly broken power-law spectrum for high-z objects
- Optimized FeII template fitting by utilizing PPXF framework
- Added UV FeII+FeIII template from Vestergaard & Wilkes (2001)
- Added Balmer continuum component
- Added equivalent width calculations
- Added additional chisquared fit statistic for outflow test
- Voigt and Gauss-Hermite line profile options, with any number of higher order moments
- Emission line list options (default and user-specified)
- Control over soft- and hard constraints
- Option for non-SDSS spectrum input
- Interpolation over metal absorption lines
- Masking of bad pixels, strong emission+absorption lines (automated), and user-defined masks
- Various bug fixes, plotting improvements
- New hypothesis testing for lines and outflows (F-test remains unchanged)
- Continuum luminosities at 1350 Å, 3000 Å, and 5100 Å.
- pathlib support
- Corner plots (corner.py) no longer supported; user should make their own corner plots with fewer free parameters
- Removed BPT diagram function; user should make BPT diagrams post processing


7.7.2 - 7.7.6
=============
- Fixed problem with FeII emission lines at the edge of the fitting region. This is done by setting the variable edge_pad=0.
- Fixed F-test NaN confidence bug
- Updated initial fitting parameters in Jupyter notebook
- Bug fixes and fixes to plots


7.7.1
=====
- MNRAS Publication Version
- Added statistical F-test for ascertaining confidence between single-Gaussian and double-Gaussian models for the outflow test. Removed the ratio-of-variance test and replaced it with a sum-of-squares of residuals ratio.
- Added "n_basinhop" to fit_options, which allows user to choose how many initial basinhopping success iterations before a solution is achieved. This can drastically reduce the basinhopping fit time, at the expense of fit quality.
- Bug fixes


7.7.0
=====
- NLS1 support; more detailed option for FeII template fitting (fwhm and voff fitting options); Lorentzian emission line profile option.
- Kovacevic et al. 2010 FeII template added, which includes a paramter for temperature.
- Relaxed wavelength requirement for outflow tests for higher-redshift targets.


7.6.0 - 7.6.8
=============
- Writing no-outflow parameters from test_outflows run to log file.
- bug fixes


7.5.0 - 7.5.3
=============
- Test outflow residual statistic replaced with f-statistic (ratio-of-variances) to compare model residuals.
- Added interpolation of bad pixels based on SDSS flagged pixels.
- bug fixes


7.4.1 - 7.4.3
=============
- Writing outflow test metrics to log file for post-fit analysis.
- Improved outflow/max-likelihood fitting using scipy.optimize.basinhopping. While basinhopping algorithm requires more runtime, it produces a significantly better fit, namely for the power-law slope parameter which never varies with the SLSQP algorithm due to the fact that it is stuck in a local minima.
- Added F-statistic (ratio of variances between no outflow and outflow model).
- Changed default outflow statistic settings.

BADASS_options.py
################################## Fit Options ################################# #######
# Fitting Parameters
fit_options={
"fit_reg"    : (4750,5100),# Fitting region; Note: Indo-US Library=(3460,9464)
"good_thresh": 0.0, # percentage of "good" pixels required in fig_reg for fit.
"mask_bad_pix": False, # mask pixels SDSS flagged as 'bad' (careful!)
"mask_emline" : False, # automatically mask lines for continuum fitting.
"mask_metal": False, # interpolate over metal absorption lines for high-z spectra
"fit_stat": "RCHI2", # fit statistic; ML = Max. Like. , LS = Least Squares, RCHI2 = reduced chi2
"n_basinhop": 2, # Number of consecutive basinhopping thresholds before solution achieved
"test_outflows": False, # only test for outflows; "fit_outflows" must be set to True!
"test_line": {"bool":False,
              "line":"NA_OIII_5007"},
"max_like_niter": 2, # number of maximum likelihood iterations
"output_pars": False, # only output free parameters of fit and stop code (diagnostic)
"cosmology": {"H0":70.0, "Om0": 0.30}, # Flat Lam-CDM Cosmology
}
################################################################################

########################### MCMC algorithm parameters ##########################
mcmc_options={
"mcmc_fit"    : False, # Perform robust fitting using emcee
"nwalkers"    : 100,  # Number of emcee walkers; min = 2 x N_parameters
"auto_stop"   : False, # Automatic stop using autocorrelation analysis
"conv_type"   : "all", # "median", "mean", "all", or (tuple) of parameters
"min_samp"    : 1000,  # min number of iterations for sampling post-convergence
"ncor_times"  : 1.0,  # number of autocorrelation times for convergence
"autocorr_tol": 10.0,  # percent tolerance between checking autocorr. times
"write_iter"  : 100,   # write/check autocorrelation times interval
"write_thresh": 100,   # iteration to start writing/checking parameters
"burn_in"     : 1500, # burn-in if max_iter is reached
"min_iter"    : 2500, # min number of iterations before stopping
"max_iter"    : 2500, # max number of MCMC iterations
}
################################################################################

############################ Fit component options #############################
comp_options={
"fit_opt_feii"     : True, # optical FeII
"fit_uv_iron"      : False, # UV Iron 
"fit_balmer"       : False, # Balmer continuum (<4000 A)
"fit_losvd"        : True, # stellar LOSVD
"fit_host"         : False, # host template
"fit_power"        : True, # AGN power-law
"fit_poly"         : False, # Add polynomial continuum component
"fit_narrow"       : True, # narrow lines
"fit_broad"        : True, # broad lines
"fit_outflow"      : True, # outflow lines
"fit_absorp"       : False, # absorption lines
"tie_line_disp"    : False, # tie line widths
"tie_line_voff"    : False, # tie line velocity offsets
"na_line_profile"  : "gaussian",     # narrow line profile
"br_line_profile"  : "gauss-hermite",     # broad line profile
"out_line_profile" : "gaussian",     # outflow line profile
"abs_line_profile" : "gaussian",     # absorption line profile
"n_moments"        : 4, # number of Gauss-Hermite moments for Gauss-Hermite line profiles
                        # must be >2 and <10 for higher-order moments (default = 4)
}
################################################################################

############################ Principal Component Analysis (PCA) options #############################
# Used for reconstructing a spectrum using templates from SDSS spectra.

pca_options = {
'do_pca'       : True,          # boolean, if True will perform principal component analysis then run BADASS on the reconstructed spectrum
'n_components' : 20,            # number of PCA components to include. Should be integer > 0 or None (to fit all possible components, a few thousand). 
'pca_masks'    : [(4760,4800),] # list of regions (wavelength, in Angstroms) to perform PCA on. If list is empty, will perform PCA over entire spectrum. 
}

################################################################################


# User defined masked regions (list of tuples)
user_mask = [
# (4850,5015),
]

########################## LOSVD Fitting & Options ##################
# For direct fitting of the stellar kinematics (stellar LOSVD), one can 
# specify a stellar template library (Indo-US, Vazdekis 2010, or eMILES).
# One can also hold velocity or dispersion constant to avoid template
# convolution during the fitting process.
################################################################################

losvd_options = {
"library"   : "IndoUS", # Options: IndoUS, Vazdekis2010, eMILES
"vel_const" : {"bool":False, "val":0.0},
"disp_const": {"bool":False, "val":250.0},
}

########################## SSP Host Galaxy Template & Options ##################
# The default is zero velocity, 100 km/s dispersion 10 Gyr template from 
# the eMILES stellar library. 
################################################################################

host_options = {
"age"       : [1.0,5.0,10.0], # Gyr; [0.09 Gyr - 14 Gyr] 
"vel_const" : {"bool":True, "val":0.0},
"disp_const": {"bool":True, "val":150.0}
}

########################### AGN power-law continuum & Options ##################
# The default is a simple power law.
################################################################################

power_options = {
"type" : "simple" # alternatively, "broken" for smoothly-broken power-law
}

########################### Polynomial Continuum Options #######################
# Disabled by default.  Options for a power series polynomial continuum, 
# additive legendre polynomial, or multiplicative polynomial to be included in 
# the fit.
################################################################################

poly_options = {
"ppoly" : {"bool": False, "order": 3}, # positive definite additive polynomial 
"apoly" : {"bool": True , "order": 3}, # Legendre additive polynomial 
"mpoly" : {"bool": False, "order": 3}, # Legendre multiplicative polynomial 
}

############################### Optical FeII options ###############################
# Below are options for fitting FeII.  For most objects, you don't need to 
# perform detailed fitting on FeII (only fit for amplitudes) use the 
# Veron-Cetty 2004 template ('VC04') (2-6 free parameters)
# However in NLS1 objects, FeII is much stronger, and sometimes more detailed 
# fitting is necessary, use the Kovacevic 2010 template 
# ('K10'; 7 free parameters).

# The options are:
# template   : VC04 (Veron-Cetty 2004) or K10 (Kovacevic 2010)
# amp_const  : constant amplitude (default False)
# disp_const : constant disp (default True)
# voff_const : constant velocity offset (default True)
# temp_const : constant temp ('K10' only)

opt_feii_options={
"opt_template"  :{"type":"VC04"}, 
"opt_amp_const" :{"bool":False, "br_opt_feii_val":1.0   , "na_opt_feii_val":1.0},
"opt_disp_const":{"bool":False, "br_opt_feii_val":3000.0, "na_opt_feii_val":500.0},
"opt_voff_const":{"bool":False, "br_opt_feii_val":0.0   , "na_opt_feii_val":0.0},
}
# or
# opt_feii_options={
# "opt_template"  :{"type":"K10"},
# "opt_amp_const" :{"bool":False,"f_feii_val":1.0,"s_feii_val":1.0,"g_feii_val":1.0,"z_feii_val":1.0},
# "opt_disp_const":{"bool":False,"opt_feii_val":1500.0},
# "opt_voff_const":{"bool":False,"opt_feii_val":0.0},
# "opt_temp_const":{"bool":True,"opt_feii_val":10000.0},
# }
################################################################################

############################### UV Iron options ################################
uv_iron_options={
"uv_amp_const"  :{"bool":False, "uv_iron_val":1.0},
"uv_disp_const" :{"bool":False, "uv_iron_val":3000.0},
"uv_voff_const" :{"bool":True,  "uv_iron_val":0.0},
"uv_legendre_p" :{"bool":False, "uv_iron_val":3},
}
################################################################################

########################### Balmer Continuum options ###########################
# For most purposes, only the ratio R, and the overall amplitude are free paramters
# but if you want to go crazy, you can fit everything.
balmer_options = {
"R_const"          :{"bool":True,  "R_val":1.0}, # ratio between balmer continuum and higher-order balmer lines
"balmer_amp_const" :{"bool":False, "balmer_amp_val":1.0}, # amplitude of overall balmer model (continuum + higher-order lines)
"balmer_disp_const":{"bool":True,  "balmer_disp_val":5000.0}, # broadening of higher-order Balmer lines
"balmer_voff_const":{"bool":True,  "balmer_voff_val":0.0}, # velocity offset of higher-order Balmer lines
"Teff_const"       :{"bool":True,  "Teff_val":15000.0}, # effective temperature
"tau_const"        :{"bool":True,  "tau_val":1.0}, # optical depth
}

################################################################################

############################### Plotting options ###############################
plot_options={
"plot_param_hist"    : False,# Plot MCMC histograms and chains for each parameter
"plot_flux_hist"     : False,# Plot MCMC hist. and chains for component fluxes
"plot_lum_hist"      : False,# Plot MCMC hist. and chains for component luminosities
"plot_eqwidth_hist"  : False, # Plot MCMC hist. and chains for equivalent widths 
"plot_HTML"          : False,# make interactive plotly HTML best-fit plot
"plot_pca"           : True, # Plot PCA reconstructed spectrum. If doing PCA, you probably want this as True
}
################################################################################

################################ Output options ################################
output_options={
"write_chain" : False, # Write MCMC chains for all paramters, fluxes, and
                         # luminosities to a FITS table We set this to false 
                         # because MCMC_chains.FITS file can become very large, 
                         # especially  if you are running multiple objects.  
                         # You only need this if you want to reconstruct chains 
                         # and histograms. 
"write_options": True,  # output restart file
"verbose"     : True,  # prints steps of fitting process in Notebook
}
################################################################################

################################### Line List ##################################
# If not specified, defaults to SDSS-QSO Emission Lines (http://classic.sdss.org/dr6/algorithms/linestable.html)
################################################################################

narrow_lines = {
    ### Region 5 (4400 Å - 5500 Å)

    "NA_H_BETA"	   :{"center":4862.691, 
                     "amp":"free", 
                     "disp":"NA_OIII_5007_DISP", 
                     "voff":"NA_OIII_5007_VOFF", 
                     "line_type":"na" ,
                     "label":r"H$\beta$"
                    },
    "NA_OIII_4960" :{"center":4960.295, 
                     "amp":"(NA_OIII_5007_AMP/2.98)", 
                     "disp":"NA_OIII_5007_DISP", 
                     "voff":"NA_OIII_5007_VOFF", 
                     "line_type":"na" ,
                     "label":r"[O III]"
                    },
    "NA_OIII_5007" :{"center":5008.240, 
                     "amp":"free", 
                     "disp":"free", 
                     "voff":"free", 
                     "line_type":"na" ,
                     "label":r"[O III]"
                    },
}

outlfow_line_profile = "G"

outflow_lines = {
    "OUT_H_BETA"    :{"center":4862.691, 
                      "amp":"OUT_OIII_5007_AMP/NA_OIII_5007_AMP*NA_H_BETA_AMP" , 
                      "disp":"OUT_OIII_5007_DISP", 
                      "voff":"OUT_OIII_5007_VOFF", 
                      "line_type":"out"
                     },
    "OUT_OIII_4960" :{"center":4960.295, 
                      "amp":"OUT_OIII_5007_AMP/2.98", 
                      "disp":"OUT_OIII_5007_DISP", 
                      "voff":"OUT_OIII_5007_VOFF", 
                      "line_type":"out"
                     },
    "OUT_OIII_5007" :{"center":5008.240, 
                      "amp":"free", 
                      "disp":"free", 
                      "voff":"free", 
                      "line_type":"out"
                     },

}

broad_lines = {
    "BR_H_BETA"   :{"center":4862.691, 
                    "amp":"free", 
                    "disp":"free", 
                    "voff":"free",
                    "h3":"free",
                    "h3_prior":{"type":"gaussian","loc":0.0,"scale":0.01},
                    "h4":"free",
                    "h4_prior":{"type":"gaussian","loc":0.0,"scale":0.01},
                    "disp_plim":(500,5000),
                    "disp_init":1000.0,
                    "line_profile":"GH",
                    "line_type":"br"
                   },
}
# Combine all line lists into single list
user_lines = {**narrow_lines, **broad_lines, **outflow_lines}#, **absorp_lines}

user_constraints = [
    # Region 5 (4400 Å - 5500 Å)
    ("NA_OIII_5007_AMP","NA_H_BETA_AMP"),# [OIII]5007 AMP > NA H-BETA AMP
    ("OUT_OIII_5007_DISP","NA_OIII_5007_DISP"), # 2nd [OIII] component DISP > 1st [OIII] component DISP
    ("NA_OIII_5007_AMP","OUT_OIII_5007_AMP"), # 2nd [OIII] component DISP > 1st [OIII] component DISP
    ]

combined_lines = {
#     "OIII_5007_COMP":["NA_OIII_5007","OUT_OIII_5007"],
#     "OIII_4960_COMP":["NA_OIII_4960","OUT_OIII_4960"],
#     "H_BETA_COMP"   :["NA_H_BETA","OUT_H_BETA"],
    }

################################################################################